File: Graphics/Testing/Cxx/MeshQuality.cxx
/*
 * Copyright 2003 Sandia Corporation.
 * Under the terms of Contract DE-AC04-94AL85000, there is a non-exclusive
 * license for use of this work by or on behalf of the
 * U.S. Government. Redistribution and use in source and binary forms, with
 * or without modification, are permitted provided that this Notice and any
 * statement of authorship are reproduced on all copies.
 */

#include "vtkMeshQuality.h"

#include "vtkDebugLeaks.h"
#include "vtkTestUtilities.h"
#include "vtkUnstructuredGrid.h"
#include "vtkUnstructuredGridReader.h"
#include "vtkFieldData.h"

int DumpQualityStats( vtkMeshQuality* iq, const char *arrayname )
{
  cout << "  cardinality: "
       << iq->GetOutput()->GetFieldData()->GetArray( arrayname )->GetComponent( 0, 4 )
       << "  , range: "
       << iq->GetOutput()->GetFieldData()->GetArray( arrayname )->GetComponent( 0, 0 )
       << "  -  "
       << iq->GetOutput()->GetFieldData()->GetArray( arrayname )->GetComponent( 0, 2 )
       << endl;

  cout << "  average: " << iq->GetOutput()->GetFieldData()->GetArray( arrayname )->GetComponent( 0, 1 )
       << "  , standard deviation: "
       << sqrt(fabs(iq->GetOutput()->GetFieldData()->GetArray( arrayname )->GetComponent( 0, 3 )))
       << endl;

  return 0;
}

int MeshQuality( int argc, char* argv[] )
{
  vtkUnstructuredGridReader* mr = vtkUnstructuredGridReader::New();
  vtkUnstructuredGrid* ug;
  vtkMeshQuality* iq = vtkMeshQuality::New();
  char* fname = vtkTestUtilities::ExpandDataFileName(argc, argv, "Data/uGridEx.vtk");

  mr->SetFileName( fname );
  mr->Update();

  ug = mr->GetOutput();
  iq->SetInput( ug );

  if ( ug->GetNumberOfCells() )
    {
    cout << endl; 
    cout << "Triangle quality of mesh" << endl;
    cout << mr->GetFileName()
         << endl;

    iq->SetTriangleQualityMeasureToEdgeRatio();
    iq->Update();
    cout << " Edge Ratio:"
         << endl;
    DumpQualityStats( iq, "Mesh Triangle Quality" );
 
    iq->SetTriangleQualityMeasureToAspectRatio();
    iq->Update();
    cout << " Aspect Ratio:"
         << endl;
    DumpQualityStats( iq, "Mesh Triangle Quality" );
    
    iq->SetTriangleQualityMeasureToRadiusRatio();
    iq->Update();
    cout << " Radius Ratio:"
         << endl;
    DumpQualityStats( iq, "Mesh Triangle Quality" );

    iq->SetTriangleQualityMeasureToAspectFrobenius();
    iq->Update();
    cout << " Frobenius Norm:"
         << endl;
    DumpQualityStats( iq, "Mesh Triangle Quality" );
    
    iq->SetTriangleQualityMeasureToMinAngle();
    iq->Update();
    cout << " Minimal Angle:"
         << endl;
    DumpQualityStats( iq, "Mesh Triangle Quality" );
 
    cout << endl; 
    cout << "Quadrilatedral quality of mesh" << endl;
    cout << mr->GetFileName()
         << endl;
    
    iq->SetQuadQualityMeasureToEdgeRatio();
    iq->Update();
    cout << " Edge Ratio:"
         << endl;
    DumpQualityStats( iq, "Mesh Quadrilateral Quality" );
    
    iq->SetQuadQualityMeasureToAspectRatio();
    iq->Update();
    cout << " Aspect Ratio:"
         << endl;
    DumpQualityStats( iq, "Mesh Quadrilateral Quality" );

    iq->SetQuadQualityMeasureToRadiusRatio();
    iq->Update();
    cout << " Radius Ratio:"
         << endl;
    DumpQualityStats( iq, "Mesh Quadrilateral Quality" );
    
    iq->SetQuadQualityMeasureToMedAspectFrobenius();
    iq->Update();
    cout << " Average Frobenius Norm:"
         << endl;
    DumpQualityStats( iq, "Mesh Quadrilateral Quality" );
    
    iq->SetQuadQualityMeasureToMaxAspectFrobenius();
    iq->Update();
    cout << " Maximal Frobenius Norm:"
         << endl;
    DumpQualityStats( iq, "Mesh Quadrilateral Quality" );
    
    iq->SetQuadQualityMeasureToMinAngle();
    iq->Update();
    cout << " Minimal Angle:"
         << endl;
    DumpQualityStats( iq, "Mesh Quadrilateral Quality" );
 
    cout << endl; 
    cout << "Tetrahedral quality of mesh" << endl;
    cout << mr->GetFileName()
         << endl;
    
    iq->SetTetQualityMeasureToEdgeRatio();
    iq->Update();
    cout << " Edge Ratio:"
         << endl;
    DumpQualityStats( iq, "Mesh Tetrahedron Quality" );
    
    iq->SetTetQualityMeasureToAspectRatio();
    iq->Update();
    cout << " Aspect Ratio:"
         << endl;
    DumpQualityStats( iq, "Mesh Tetrahedron Quality" );
    
    iq->SetTetQualityMeasureToRadiusRatio();
    iq->Update();
    cout << " Radius Ratio:"
         << endl;
    DumpQualityStats( iq, "Mesh Tetrahedron Quality" );
    
    iq->SetTetQualityMeasureToAspectFrobenius();
    iq->Update();
    cout << " Frobenius Norm:"
         << endl;
    DumpQualityStats( iq, "Mesh Tetrahedron Quality" );
    
    iq->SetTetQualityMeasureToMinAngle();
    iq->Update();
    cout << " Minimal Dihedral Angle:"
         << endl;
    DumpQualityStats( iq, "Mesh Tetrahedron Quality" );

    iq->SetTetQualityMeasureToCollapseRatio();
    iq->Update();
    cout << " Collapse Ratio:"
         << endl;
    DumpQualityStats( iq, "Mesh Tetrahedron Quality" );
    cout << endl; 

    cout << "Hexahedral quality of mesh" << endl;
    cout << mr->GetFileName()
         << endl;
    
    iq->SetHexQualityMeasureToEdgeRatio();
    iq->Update();
    cout << " Edge Ratio:"
         << endl;
    DumpQualityStats( iq, "Mesh Hexahedron Quality" );
    cout << endl; 
    }

  iq->Delete();
  mr->Delete();
  delete [] fname;

  return 0;
}

File: Filters/Verdict/Testing/Python/meshQuality.py
#!/usr/bin/env python

import sys
import math

for i in range(0, len(sys.argv)):
    if sys.argv[i] == '-A' and i < len(sys.argv)-1:
        sys.path = sys.path + [sys.argv[i+1]]

import vtk
from vtk.util.misc import vtkGetDataRoot

filename = vtkGetDataRoot() + '/Data/tetraMesh.vtk'

def DumpQualityStats(iq, arrayname):
    print "  cardinality:",
    print iq.GetOutput().GetFieldData().GetArray(arrayname).GetComponent(0, 4),
    print " ,  range:",
    print iq.GetOutput().GetFieldData().GetArray(arrayname).GetComponent(0, 0),
    print " - ",
    print iq.GetOutput().GetFieldData().GetArray(arrayname).GetComponent(0, 2)

    print "  average:",
    print iq.GetOutput().GetFieldData().GetArray(arrayname).GetComponent(0, 1),
    print " ,  standard deviation:",
    print math.sqrt(abs(iq.GetOutput().GetFieldData().GetArray(arrayname).GetComponent(0, 3)))

mr = vtk.vtkUnstructuredGridReader()
mr.SetFileName(filename)

iq = vtk.vtkMeshQuality()
iq.SetInputConnection(mr.GetOutputPort())
iq.VolumeOn()
iq.RatioOn()
iq.Update()

ug = iq.GetOutput()
iq.SetInputConnection(mr.GetOutputPort())

if ug.GetNumberOfCells():
    print
    print "Triangle quality of mesh"
    print mr.GetFileName()

    iq.SetTriangleQualityMeasureToEdgeRatio()
    iq.Update()
    print " Edge Ratio:"

    DumpQualityStats(iq, "Mesh Triangle Quality")

    iq.SetTriangleQualityMeasureToAspectRatio()
    iq.Update()
    print " Aspect Ratio:"

    DumpQualityStats(iq, "Mesh Triangle Quality")

    iq.SetTriangleQualityMeasureToRadiusRatio()
    iq.Update()
    print " Radius Ratio:"

    DumpQualityStats(iq, "Mesh Triangle Quality")

    iq.SetTriangleQualityMeasureToAspectFrobenius()
    iq.Update()
    print " Frobenius Norm:"

    DumpQualityStats(iq, "Mesh Triangle Quality")

    iq.SetTriangleQualityMeasureToMinAngle()
    iq.Update()
    print " Minimal Angle:"

    DumpQualityStats(iq, "Mesh Triangle Quality")

    print "Quadrilatedral quality of mesh"
    print mr.GetFileName()

    iq.SetQuadQualityMeasureToEdgeRatio()
    iq.Update()
    print " Edge Ratio:"

    DumpQualityStats(iq, "Mesh Quadrilateral Quality")

    iq.SetQuadQualityMeasureToAspectRatio()
    iq.Update()
    print " Aspect Ratio:"

    DumpQualityStats(iq, "Mesh Quadrilateral Quality")

    iq.SetQuadQualityMeasureToRadiusRatio()
    iq.Update()
    print " Radius Ratio:"

    DumpQualityStats(iq, "Mesh Quadrilateral Quality")

    iq.SetQuadQualityMeasureToMedAspectFrobenius()
    iq.Update()
    print " Average Frobenius Norm:"

    DumpQualityStats(iq, "Mesh Quadrilateral Quality")

    iq.SetQuadQualityMeasureToMaxAspectFrobenius()
    iq.Update()
    print " Maximal Frobenius Norm:"

    DumpQualityStats(iq, "Mesh Quadrilateral Quality")

    iq.SetQuadQualityMeasureToMinAngle()
    iq.Update()
    print " Minimal Angle:"

    DumpQualityStats(iq, "Mesh Quadrilateral Quality")

    print "Tetrahedral quality of mesh"
    print mr.GetFileName()

    iq.SetTetQualityMeasureToEdgeRatio()
    iq.Update()
    print " Edge Ratio:"

    DumpQualityStats(iq, "Mesh Tetrahedron Quality")

    iq.SetTetQualityMeasureToAspectRatio()
    iq.Update()
    print " Aspect Ratio:"

    DumpQualityStats(iq, "Mesh Tetrahedron Quality")

    iq.SetTetQualityMeasureToRadiusRatio()
    iq.Update()
    print " Radius Ratio:"

    DumpQualityStats(iq, "Mesh Tetrahedron Quality")

    iq.SetTetQualityMeasureToAspectFrobenius()
    iq.Update()
    print " Frobenius Norm:"

    DumpQualityStats(iq, "Mesh Tetrahedron Quality")

    iq.SetTetQualityMeasureToMinAngle()
    iq.Update()
    print " Minimal Dihedral Angle:"

    DumpQualityStats(iq, "Mesh Tetrahedron Quality")

    iq.SetTetQualityMeasureToCollapseRatio()
    iq.Update()
    print " Collapse Ratio:"

    DumpQualityStats(iq, "Mesh Tetrahedron Quality")

    print "Hexahedral quality of mesh"
    print mr.GetFileName()

    iq.SetHexQualityMeasureToEdgeRatio()
    iq.Update()
    print " Edge Ratio:"

    DumpQualityStats(iq, "Mesh Hexahedron Quality")
